### PR TITLE
Add missing FS:USER functions

### PR DESCRIPTION
**Pull request by PabloMK7**
Adds product info and save data secure value related functions.

This PR, #7050 and https://github.com/citra-emu/citra/pull/7035 allow citra to go online using LLE friends module.

- - -
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/7051)

### CODE REVIEW

File: src/core/hle/service/fs/fs_user.cpp [hunk 1]
```diff
@@ -816,7 +837,79 @@ void FS_USER::GetSaveDataSecureValue(Kernel::HLERequestContext& ctx) {
     rb.Push<u64>(0);      // the secure value
 }
 
-void FS_USER::Register(u32 process_id, u64 program_id, const std::string& filepath) {
+void FS_USER::SetThisSaveDataSecureValue(Kernel::HLERequestContext& ctx) {
+    IPC::RequestParser rp(ctx);
+    u32 secure_value_slot = rp.Pop<u32>();
```
Comment (GPUCode) [inline src/core/hle/service/fs/fs_user.cpp:842]:
```suggestion
    const u32 secure_value_slot = rp.Pop<u32>();
```
ditto for other rp.Pop calls

File: src/core/hle/service/fs/fs_user.cpp [hunk 2]
```diff
@@ -816,7 +837,79 @@ void FS_USER::GetSaveDataSecureValue(Kernel::HLERequestContext& ctx) {
...
+
+    IPC::RequestBuilder rb = rp.MakeBuilder(1, 0);
+
+    rb.Push(RESULT_SUCCESS);
+}
+
+void FS_USER::GetThisSaveDataSecureValue(Kernel::HLERequestContext& ctx) {
+    IPC::RequestParser rp(ctx);
+
+    u32 secure_value_slot = rp.Pop<u32>();
```
Comment (GPUCode) [inline src/core/hle/service/fs/fs_user.cpp:858]:
```suggestion
    IPC::RequestParser rp(ctx);
    u32 secure_value_slot = rp.Pop<u32>();
```
ditto for the other cases of this

**Review by GPUCode — APPROVED** [review]

**Review by Steveice10 — APPROVED** [review]

File: src/core/hle/service/fs/fs_user.cpp [hunk 3]
```diff
@@ -708,8 +708,20 @@ void FS_USER::GetProgramLaunchInfo(Kernel::HLERequestContext& ctx) {
...
+
+    // Always report the launched program mediatype is SD if the friends module is requesting this
+    // information and the media type is game card. Otherwise, friends will append a "romid" field
+    // to the NASC request with a cartridge unique identifier. Using a dump of a game card and the
+    // game card itself at the same time online is known to have caused issues in the past.
+    auto process = ctx.ClientThread()->owner_process.lock();
+    if (process && process->codeset->name == "friends" &&
+        program_info.media_type == MediaType::GameCard) {
+        program_info.media_type = MediaType::SDMC;
+    }
```
Comment (DaniElectra) [inline src/core/hle/service/fs/fs_user.cpp:721]: There are some catridge-only title IDs that use online functionality. Do we know if those have issues when setting the media type to SD?

Comment (PabloMK7) [reply src/core/hle/service/fs/fs_user.cpp:721]: That's a good question. Pretty sure we would know as many people backup and install their games as .cia

File: src/core/hle/service/fs/fs_user.cpp [hunk 4]
```diff
@@ -670,6 +670,27 @@ void FS_USER::GetFormatInfo(Kernel::HLERequestContext& ctx) {
     rb.Push<bool>(format_info->duplicate_data != 0);
 }
 
+void FS_USER::GetProductInfo(Kernel::HLERequestContext& ctx) {
+    IPC::RequestParser rp(ctx);
+
+    u32 process_id = rp.Pop<u32>();
+
+    LOG_DEBUG(Service_FS, "process_id={}", process_id);
```
Comment (FearlessTobi) [inline src/core/hle/service/fs/fs_user.cpp:678]:
```suggestion
    LOG_DEBUG(Service_FS, "called, process_id={}", process_id);
```

File: src/core/hle/service/fs/fs_user.cpp [hunk 5]
```diff
@@ -670,6 +670,27 @@ void FS_USER::GetFormatInfo(Kernel::HLERequestContext& ctx) {
...
+
+    IPC::RequestBuilder rb = rp.MakeBuilder(6, 0);
+
+    auto product_info = GetProductInfo(process_id);
+    if (!product_info.has_value()) {
+        rb.Push(ResultCode(FileSys::ErrCodes::ArchiveNotMounted, ErrorModule::FS,
+                           ErrorSummary::NotFound, ErrorLevel::Status));
+        rb.Skip(5, false);
```
Comment (FearlessTobi) [inline src/core/hle/service/fs/fs_user.cpp:686]: is this needed?

File: src/core/hle/service/fs/fs_user.cpp [hunk 6]
```diff
@@ -670,6 +670,27 @@ void FS_USER::GetFormatInfo(Kernel::HLERequestContext& ctx) {
...
+void FS_USER::GetProductInfo(Kernel::HLERequestContext& ctx) {
+    IPC::RequestParser rp(ctx);
+
+    u32 process_id = rp.Pop<u32>();
+
+    LOG_DEBUG(Service_FS, "process_id={}", process_id);
+
+    IPC::RequestBuilder rb = rp.MakeBuilder(6, 0);
+
+    auto product_info = GetProductInfo(process_id);
```
Comment (FearlessTobi) [inline src/core/hle/service/fs/fs_user.cpp:682]:
```suggestion
    const auto product_info = GetProductInfo(process_id);
```

File: src/core/hle/service/fs/fs_user.cpp [hunk 7]
```diff
@@ -1002,4 +1121,4 @@ void InstallInterfaces(Core::System& system) {
     auto& service_manager = system.ServiceManager();
     std::make_shared<FS_USER>(system)->InstallAsService(service_manager);
 }
-} // namespace Service::FS
+} // namespace Service::FS
```
Comment (FearlessTobi) [inline src/core/hle/service/fs/fs_user.cpp:1124]: newline

File: src/core/hle/service/fs/fs_user.h [hunk 8]
```diff
@@ -4,8 +4,10 @@
 
 #pragma once
 
+#include <optional>
 #include <unordered_map>
 #include <boost/serialization/base_object.hpp>
+#include "archive.h"
```
Comment (FearlessTobi) [inline src/core/hle/service/fs/fs_user.h:10]: please use the full path for includes

File: src/core/hle/service/fs/fs_user.h [hunk 9]
```diff
@@ -47,12 +49,22 @@ class FS_USER final : public ServiceFramework<FS_USER, ClientSlot> {
...
+    // loader and pm, which we HLEed, we can just directly use it here
+    void RegisterProgramInfo(u32 process_id, u64 program_id, const std::string& filepath);
 
     std::string GetCurrentGamecardPath() const;
 
+    struct ProductInfo {
+        std::array<u8, 0x10> product_code;
+        u16_le maker_code;
+        u16_le remaster_version;
+    };
```
Comment (FearlessTobi) [inline src/core/hle/service/fs/fs_user.h:62]: should have an assert for its size

File: src/core/hle/service/fs/fs_user.h [hunk 10]
```diff
@@ -643,4 +719,4 @@ void InstallInterfaces(Core::System& system);
 
 SERVICE_CONSTRUCT(Service::FS::FS_USER)
 BOOST_CLASS_EXPORT_KEY(Service::FS::FS_USER)
-BOOST_CLASS_EXPORT_KEY(Service::FS::ClientSlot)
+BOOST_CLASS_EXPORT_KEY(Service::FS::ClientSlot)
```
Comment (FearlessTobi) [inline src/core/hle/service/fs/fs_user.h:722]: newline

File: src/core/loader/ncch.cpp [hunk 11]
```diff
@@ -174,7 +174,15 @@ ResultStatus AppLoader_NCCH::LoadExec(std::shared_ptr<Kernel::Process>& process)
         auto fs_user =
             Core::System::GetInstance().ServiceManager().GetService<Service::FS::FS_USER>(
                 "fs:USER");
-        fs_user->Register(process->process_id, process->codeset->program_id, filepath);
+        fs_user->RegisterProgramInfo(process->process_id, process->codeset->program_id, filepath);
+
+        Service::FS::FS_USER::ProductInfo product_info{};
+        memcpy(product_info.product_code.data(), overlay_ncch->ncch_header.product_code,
```
Comment (FearlessTobi) [inline src/core/loader/ncch.cpp:180]:
```suggestion
        std::memcpy(product_info.product_code.data(), overlay_ncch->ncch_header.product_code,
```

File: src/core/loader/ncch.cpp [hunk 12]
```diff
@@ -174,7 +174,15 @@ ResultStatus AppLoader_NCCH::LoadExec(std::shared_ptr<Kernel::Process>& process)
...
+               product_info.product_code.size());
+        std::memcpy(&product_info.remaster_version,
+                    overlay_ncch->exheader_header.codeset_info.flags.remaster_version, 2);
```
Comment (FearlessTobi) [inline src/core/loader/ncch.cpp:183]: magic number, can probably be a `sizeof`

File: src/core/hle/service/fs/fs_user.cpp (hunk 3 above)
Comment (PabloMK7) [reply src/core/hle/service/fs/fs_user.cpp:721]: Researched a bit and couldn't find any issues of people with the cartridge exclusive games converted as cia.

**Pull request merged**